modules/objdetect/perf/perf_qrcode_pipeline.cpp

```diff
@@ -0,0 +1,111 @@
+    const std::string root = "cv/qrcode/";
+
+    std::string image_path = findDataFile(root + name_current_image);
+    Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+    std::vector< Point > corners;
+    TEST_CYCLE() ASSERT_TRUE(detectQRCode(src, corners));
+    SANITY_CHECK(corners);
+}
+
+#ifdef HAVE_QUIRC
+PERF_TEST_P_(Perf_Objdetect_QRCode, decode)
+{
+    const std::string name_current_image = GetParam();
+    const std::string root = "cv/qrcode/";
+
+    std::string image_path = findDataFile(root + name_current_image);
+    Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+    std::vector< Point > corners;
+    std::string decoded_info;
+    ASSERT_TRUE(detectQRCode(src, corners));
+    TEST_CYCLE() ASSERT_TRUE(decodeQRCode(src, corners, decoded_info, straight_barcode));
+
+    std::vector<uint8_t> decoded_info_uint8_t(decoded_info.begin(), decoded_info.end());
+    SANITY_CHECK(decoded_info_uint8_t);
+    SANITY_CHECK(straight_barcode);
+
+}
+#endif
+
+INSTANTIATE_TEST_CASE_P(/*nothing*/, Perf_Objdetect_QRCode,
+    ::testing::Values(
+        "version_1_down.jpg", "version_1_left.jpg", "version_1_right.jpg", "version_1_up.jpg", "version_1_top.jpg",
+        "version_5_down.jpg", "version_5_left.jpg", "version_5_right.jpg", "version_5_up.jpg", "version_5_top.jpg",
+        "russian.jpg", "kanji.jpg", "link_github_ocv.jpg", "link_ocv.jpg", "link_wiki_cv.jpg"
+    )
+);
+
+Mat getNotQRCodeMat(std::string type_gen)
+{
+    Mat image(256, 256, CV_8UC1);
+    if (type_gen == "not_random") { image = Mat::zeros(256, 256, CV_8UC1); }
+    if (type_gen == "random")
+    {
+        RNG rng;
+        for(int i = 0; i < image.rows; i++)
+        {
+            for(int j = 0; j < image.cols; j++)
+            {
+                image.at<uint8_t>(i,j) = static_cast<uint8_t>(rng.uniform(0, 255));
+            }
+        }
+    }
+    return image;
+}
+
+typedef ::perf::TestBaseWithParam< std::string > Perf_Objdetect_Not_QRCode;
+
+PERF_TEST_P_(Perf_Objdetect_Not_QRCode, detect)
+{
+    Mat straight_barcode;
+    std::vector<Point> corners;
+    std::string decoded_info;
+    Mat not_qr_code = getNotQRCodeMat(GetParam());
+
+    TEST_CYCLE() ASSERT_FALSE(detectQRCode(not_qr_code, corners));
+    SANITY_CHECK_NOTHING();
+}
+
+#ifdef HAVE_QUIRC
+PERF_TEST_P_(Perf_Objdetect_Not_QRCode, decode)
+{
+    const std::string path_current_image = "cv/qrcode/version_1_down.jpg";
+
+    std::string image_path = findDataFile(path_current_image);
+    Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+    std::vector< Point > corners;
+    std::string decoded_info;
+    ASSERT_TRUE(detectQRCode(src, corners));
+
+    Mat not_qr_code = getNotQRCodeMat(GetParam());
```

Comment:
It is better to fill corners in this generation function instead of using some image (pre-defined points should be fine).

Test inputs must be clear as much as possible and **don't expect correctness of other functions** (int this case - correctness of "detection" `detectQRCode()`).

Changes in other functions should not change the test result.